### PR TITLE
Remove experimental torch.distributed._dist2 API

### PR DESCRIPTION
**Pull request by d4l3k**
Removes the experimental object-oriented distributed API `torch.distributed._dist2` along with its test and documentation.

The module's docstring marked it as experimental and subject to deletion entirely, and nothing in the tree imports it outside its own test. This removes:

- `torch/distributed/_dist2.py` (implementation)
- `test/distributed/test_dist2.py` (tests)
- `docs/source/distributed._dist2.md` and its toctree entry in `docs/source/distributed.md` (docs)

The shared C++ bindings it used (`_current_process_group` / `_set_process_group`) are left untouched.

A [GitHub code search](https://github.com/search?q=%2F%28from%7Cimport%29.*distributed.*_dist2%2F&type=code) finds only a single external user ([cocoa-org/miles](https://github.com/cocoa-org/miles/blob/ca11e47a8cf4522aa7bc95cce2f41568e5c4162f/miles/backends/training_utils/loss.py#L997)). Since the API is experimental and explicitly marked as subject to deletion, it is fine to remove.

## Test Plan

```
git grep -n "distributed._dist2" -- ':!build/*'   # no matches
python -c "import torch.distributed"              # imports cleanly
python -c "import torch.distributed._dist2"        # ModuleNotFoundError
```

### CODE REVIEW

**Review by amirafzali — APPROVED** [review]
lgtm 🧹

**Review by kapilsh — APPROVED** [review]

**Review by dcci — APPROVED** [review]
yay

**Review by fduwjj — APPROVED** [review]

**Comment by fduwjj** [comment]
@pytorch merge -i

**Comment by fduwjj** [comment]
I confirmed with @fegin that the torch titan test is flaky..
